File: d05.cc
#include <iostream>
#include <vector>

#include "d05d.cc"

using namespace std;

int p1(vector<int> maze){
  int step = 0;
  for (vector<int>::iterator it = maze.begin(); it >= maze.begin() && it < maze.end(); ++step) {
    it += (*it)++;
  }
  return step;
}

int p2(vector<int> maze){
  int step = 0;
  for (vector<int>::iterator it = maze.begin(); it >= maze.begin() && it < maze.end(); ++step) {
    it += (*it < 3 ? (*it)++ : (*it)--);
  }
  return step;
}

int main(){
  cout << p1(initial_maze) << endl;
  cout << p2(initial_maze) << endl;
}

File: d05.rb
def p1(maze)
  step = 0
  idx = 0
  while idx < maze.size && idx >= 0 do
    maze[idx] += 1
    idx += maze[idx] - 1
    step += 1
  end
  step
end

def p2(maze)
  step = 0
  idx = 0
  while idx < maze.size && idx >= 0 do
    old_v = maze[idx]
    maze[idx] += maze[idx] >= 3 ? -1 : 1
    idx += old_v
    step += 1
  end
  step
end
